### ISSUE TITLE
JavaScript: Move 'Linting' lesson (again) -OR- Add a short intro to NPM(?)

### ISSUE THREAD

**Issue by takinabradley** [TheOdinProject/curriculum#24987]
### Describe your suggestion

I've noticed that since moving the Linting lesson in the JavaScript course, questions surrounding how and when to install ESLint have increased quite a bit in the Discord. The lesson was moved after a conversation about whether ESLint/Prettier should be more strongly recommended and recommended earlier in the curriculum:
https://discord.com/channels/505093832157691914/1049683144900628512

It seems to be the uptick is probably due to:
1. More people are making to this lesson, quicker (It was moved from AFTER  the 'Organizing Your JS Code' section to BEFORE)
2. Learners not having any experience installing NPM packages in their projects yet (A real introduction to NPM isn't presented until the 'ES6 Modules' lesson)

I feel a quick introduction to NPM in/before the 'Linting' lesson might help learners and lead to a reduction in questions asked about installing/using ESLint. Another option would be to move the lesson back to after the 'Organizing Your JS Code' section while still encouraging it's use throughout the remainder of the course.

### Path

Node / JS

### Lesson Url

https://www.theodinproject.com/lessons/node-path-javascript-linting

### Checks

- [X] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/.github/blob/main/CONTRIBUTING.md)
- [X] Would you like to work on this issue?

### (Optional) Discord Name

GrenadeShit#1112

### (Optional) Additional Comments

Relevant Commits:
https://github.com/TheOdinProject/curriculum/commit/3e5ffe2aabecc0123008513bf11864b31648a868
https://github.com/TheOdinProject/theodinproject/commit/f1b23d433b4234a6a5d7e099342a3bcb0748359b

Suggestions-Bugs Discord Thread:
https://discord.com/channels/505093832157691914/1055733983616438323

Common types of questions I've seen I feel are related:
- Do prettier and ESLint is needed to installed in every project?
- ESLint stopped working/works in some files but not others (Similar to above, learner only installed in a single repo)
- `npm init @eslint/config` not working (npm not initialized)
- Where do I run `npm init @eslint/config`?

**Comment by takinabradley** [comment]
Oh yeah, and I'd like to add that this lesson has been moved in 'Full Stack JS' path, but not in 'Full Stack Ruby on Rails'

**Comment by scheals** [comment]
As a fellow learner I find this issue rather important, so ++ from me on having this being worked on - there was pretty much an explosion of confusion after this move.

**Comment by maximdrz** [comment]
Speaking from the perspective of someone who this issue affected - I completely agree that the integrity of the course fell apart here. Adding an intro to NPM is a great idea. I don't really know what npm really is yet.

**Comment by Lofty-Brambles** [comment]
Here's the pain-point that I don't think I see. As a learner, they don't get introduced to template repos, or config files or even a simple introduction to how the package.json works. 

It's a lesson explaining linting- cool- And then it's all over into the dark world of executing commands and installing packages. Considering we don't know what the learner may do, they may jump straight into docs and start doing global installs. They may start with different eslint/prettier configs for different repos due to creating configs multiple times. They may be confused with the config options to select. Overall, there's so many different ways to do these things, and a learner wouldn't be expected to know the tradeoffs. They may even be confused to know what the config files does. And let's not mention the number of times someone's missed eslint-config-prettier. 

And honestly, stuffing all that into one lesson, which'd have to be in the introduction bits, would be detrimental to getting started, heck, webpack is confusing as it is, node + npm would be harder to understand.

Here's something that @ManonLef mentioned:

> Yeah so as I see it a rewrite could look like this;
> Confine that lesson to prettier,
> Explain some base settings so your code will be formatted nicely (since that’s the reason it got pushed forward in the curriculum), explain to do so manually only and not use it on any collaboration projects and why. 
> Then move eslint to a spot in the curriculum where it’s more appropriate based on the knowledge present there?

Here's what I'd propose then:
 
> Move it down __after the restaurant page project__. Once webpack is done, the learner knows enough of node and npm to be able to understand what these files do, and they have practice with it. __Keep a mention of the prettier extension at the introduction lesson, as a way to save the helpers from reading weird code__.

**Comment by xandora** [comment]
It's not quite true that learners don't have any experience with npm packages before the linting lesson. They are introduced to them in the JavaScript basics section through the javascript-exercises.

**Comment by ManonLef** [comment]
I'm leaving this here in case it's decided this lesson needs a rewrite. This step-by-step seems to work for me. I spent some serious time today to extract all the steps needed from the various sources. With some discord help to iron out some things I ended up with a checklist that I eventually tested successfully and decided to edit here and there to link to for other users getting lost in the sea of information.

[Discord conversation and resulting checklist with MDFrank](https://discord.com/channels/505093832157691914/514204667245363200/1073359687531106324)

The resulting checklist with steps that worked for me is this:

# Workflow
Note that all the steps are done from the repo you're working in. We are only setting this up on a per project basis.

## Setting up prettier
1) `npm install --save-dev --save-exact prettier` in the repo you're working on 
2) `echo {}> .prettierrc.json` to create a file that is needed to recognise this repo uses prettier
3) create a `.prettierignore` file with `touch .prettierignore` and put files to be ignored in it based on your ESLintignore(? only came across this in the prettier docs so far) and/or gitignore file. See note below
4) install Prettier extension in VSCode to be able to format using a shortcut or set it up to run on save etc.

Now to manually use prettier for formatting you can use the shortcut `shift+alt+f` or use the command `npx prettier --write <filename>` 

Note for step 3:
My file is set up like this, not sure if this is correct yet, so need to confirm later:
```node_modules
package.lock.json
build
.DS_Store
CVS
Thumbs.db
```

## Setting up ESLint
5. run `npm init` or `npm init --yes` (the last to answer the prompts with yes automatically) in the repo/project I'll be working on. This is to get a package.json file in there

_6. `npm install --save-dev eslint`  to install the ESLint package in this specific repo
7. `./node_modules/.bin/eslint --init` to activate ESLint in my project and creating an ESLint configuration file*_

**OR**

_6/7)`npm init @eslint/config` to hit two birds with one stone (intitialize and creating the config file)_

8) configure the resulting eslintrc.js file first by the prompts and/or review https://www.digitalocean.com/community/tutorials/linting-and-formatting-with-eslint-in-vs-code for prompt decisions and later adjusting of config 
9) install the ESLint VSCode extension for visual clues during programming and maybe even configure it to do it's magic on each save automatically, also explained on the above DigitalOcean lesson

## Setting up eslint-config-prettier
10) `npm install --save-dev eslint-config-prettier` in the same repo to make ESLint and Prettier work nicely together without conflict
11)   add "prettier" to the "extends" array in your .eslintrc.* file at the end of the file 
example:
``` 
 "extends": [
        "airbnb-base",
        "prettier"
    ],
```
12) `npx eslint-config-prettier path/to/main.js` to see if there are conflicts

To see all the errors and warnings for ESLint you can use the command `npx eslint <filename>` but they're also shown in the VSCode Problems section 

13) set up ESLint to automatically fix syntax and formatting issues upon save by following Step 4 and 5 of this article: https://www.digitalocean.com/community/tutorials/linting-and-formatting-with-eslint-in-vs-code

**Comment by Asartea** [comment]
So, as the person who is probably to blame for this change, I largely agree with moving it back. At the time, it seemed like a wise decision, but its clearly causing far too much confusion for the added net worth of having it there. I do think we have started conflated a few different issues here:
- Should the lesson be where it is?
- What should the content of the lesson be?
- Should we do more with npm in general? 

## Should the lesson be where it is?
On this I think the consensus, both here and in what I've seen on Discord, is pretty clear: no. Its useful yes, but its causing confusion among learner because it uses a lot of high level concepts that they haven't been introduced to yet, and runs the risk of letting the learner make all kind of sneaky mistakes because they aren't yet at a level where they have the experience to fully understand how to use these tools

## What should the content of the lesson be?
Right now the linting lesson is a bit of a mess. It essentially tries to introduce three concepts:
- Style guides
- Code formatters to enforce those guides
- ESLint as a code analysis and checker tool
This leads to a lesson that is full of high level overview concepts, but misses a lot of practical examples and stuff on how to actually do what is being described in the lessons, leading to confusion about how any of this works, combined with the lack of knowledge on the high level concepts of using npm

## Should we do more with npm in general?
This last one I've seen floating around a bit, which is that we don't do a lot with npm in general, despite it being (arguably) the single most important tool in modern JavaScript development. It features in <https://www.theodinproject.com/lessons/node-path-javascript-es6-modules> as background reading for the concept of webpack, but is never truly explained.

Given all of this, I would propose the following:
- We split the current linting lesson in two pieces: Style guides with Prettier, and ESLint for linting
    - The Prettier + Style guides lesson takes the place of the current linting lesson. From my understanding prettier can be run without any local modules; alternatively we can provide step by step instructions for installing just it, without digging into the core concepts.
    - The now detached ESLint lesson moves behind the restaurant project, as suggested by @Lofty-Brambles. By that point learners should have gained enough understanding of npm and related tools to use it effectively. It can also be expanded with a more detailed guide on how to install it (@ManonLef has written most of that one reply earlier), and any issues that might arise should be solvable by the learner at that point
- On the topic of npm, I propose we split out the current npm + yarn parts from https://www.theodinproject.com/lessons/node-path-javascript-es6-modules and move them into their own lesson one lesson before that lesson, with a expanded overview and focus on npm and how to use it

This should hopefully alleviate the current confusion by more clearly separating the different concepts, and also moving them into places more inline with the level of knowledge we expect for them.

**Comment by CouchofTomato** [comment]
@TheOdinProject/javascript can someone pick this up to finalise please.

**Comment by wise-king-sullyman** [comment]
My gut feeling is that we should probably move the lesson back regardless of if we make changes to it or how we cover topics like linting/formatting/npm/packaging/ignore files/etc.

I'm actually working on an initiative that _might_ make installing and configuring ESLint and Prettier a non-issue so I would advise that we hold off on work in that direction for the moment. I don't want to give too much away as it may not work out, but it's in the template repo kind of space that was mentioned above.

@takinabradley I see that you marked that you would like to work on this issue, are you still interested in doing that? The moving of the lesson would need to happen in [our site repo](https://github.com/TheOdinProject/theodinproject) rather than this one if you're not familiar with that process.

**Comment by takinabradley** [comment]
@wise-king-sullyman I'm not familiar with the guidelines and stuff for making pull requests on the site repo, but I'll look into it soon, probably tomorrow.

For the moment, do we want resolve this issue simply by moving the lesson back to where it was previously? Now that I'm a bit further in the curriculum, I think I'd be willing to work on some changes to the lesson as well if we can come to some kind of agreement on how we'd like to change it/split it up.

**Comment by wise-king-sullyman** [comment]
@takinabradley if after reviewing our contribution guides and such you're still not sure how to go about it just let me know, happy to help you out. The site can be tricky to get a handle on if you've never done anything with it before 🙂 

And yeah for the moment I would say we just move the lesson back, then focus on improving the lesson in the not distant future.

**Comment by takinabradley** [comment]
Looks like I got it done with minimal issues. The biggest ones were figuring how to set up stuff for a non-debian environment. 
I didn't quite expect so many system installs! :sob:

**Comment by wise-king-sullyman** [comment]
Yeah our instructions are definitely tailored for Ubuntu. I went through the hassle of getting it running on Fedora once... but I don't plan to do that again. If you'd like to document what all you had to do for your OS (I assume Mac?) that is something that would probably be helpful for others in the future.

Glad you got it handled without too much trouble though! 🚀
